plugins/templates-maven-plugin/src/main/java/com/google/cloud/teleport/plugin/maven/TemplatesStageMojo.java

```diff
@@ -402,8 +452,78 @@ protected String stageFlexTemplate(
               + "\n"
               + new String(process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8));
     }
+  }
 
-    LOG.info("Flex Template was staged! {}", templatePath);
-    return templatePath;
+  private void stageFlexPythonTemplate(
+      TemplateDefinitions definition,
+      String currentTemplateName,
+      String imagePath,
+      File metadataFile,
+      String containerName,
+      String templatePath)
+      throws IOException, InterruptedException, TemplateException {
+
+    if (dockerfilePath == null) {
+      PythonDockerfileGenerator.generateDockerfile(
+          basePythonContainerImage, containerName, outputClassesDirectory);
+    }
+    stageUsingDockerfile(imagePath);
+
+    String[] flexTemplateBuildCmd =
+        new String[] {
+          "gcloud",
+          "dataflow",
+          "flex-template",
+          "build",
+          templatePath,
+          "--image",
+          imagePath,
+          "--project",
+          projectId,
+          "--sdk-language",
+          definition.getTemplateAnnotation().type().name(),
+          "--metadata-file",
+          outputClassesDirectory.getAbsolutePath() + "/" + metadataFile.getName(),
+          "--additional-user-labels",
+          "goog-dataflow-provided-template-name="
+              + currentTemplateName.toLowerCase()
+              + ",goog-dataflow-provided-template-version="
+              + TemplateDefinitionsParser.parseVersion(stagePrefix)
+              + ",goog-dataflow-provided-template-type=flex"
+        };
+    LOG.info("Running: {}", String.join(" ", flexTemplateBuildCmd));
+
+    Process process = Runtime.getRuntime().exec(flexTemplateBuildCmd);
+
+    if (process.waitFor() != 0) {
+      throw new RuntimeException(
+          "Error building template using gcloud. Please make sure it is up to date. "
+              + new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8)
+              + "\n"
+              + new String(process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8));
+    }
+  }
+
+  private void stageUsingDockerfile(String imagePath) throws IOException, InterruptedException {
+    String[] gcloudBuildsCmd = new String[] {"gcloud", "builds", "submit", "--tag", imagePath};
+    LOG.info("Running: {}", String.join(" ", gcloudBuildsCmd));
+
+    Process process =
+        Runtime.getRuntime()
+            .exec(
+                gcloudBuildsCmd,
+                null,
+                new File(
+                    dockerfilePath == null
+                        ? outputClassesDirectory.getAbsolutePath()
+                        : dockerfilePath));
+
+    if (process.waitFor() != 0) {
+      throw new RuntimeException(
+          "Error staging container image using gcloud."
+              + new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8)
+              + "\n"
+              + new String(process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8));
+    }
```

Comment:
Not entirely sure, but this got stuck to me in a test.

```
04:38:47 [INFO] Running: gcloud builds submit --tag gcr.io/{PROJECT}/2023_08_16-03/streaming-llm
06:02:36 Build timed out (after 180 minutes). Marking the build as aborted.
```

My best bet would be that we are not giving `--project` and my instance was configured elsewhere and couldn't make progress, but I'll have to test it first.

Just in case, can you provide the projectId here to avoid doing in some other place?

Reply:
I'm a little confused on what you mean, but the imagePath/project name is coming from here (which I didn't make changes to): https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/1a77a7b2e534911510876734cb7a218baa981032/plugins/templates-maven-plugin/src/main/java/com/google/cloud/teleport/plugin/maven/TemplatesStageMojo.java#L342

Could you post the full command of the test you are trying to run?

Reply:
```
export PROJECT=cloud-teleport-testing                                                    
export BUCKET_NAME=cloud-teleport-testing-it

mvn clean install -pl plugins/templates-maven-plugin -am -DskipTests -DskipShade

mvn clean package -PtemplatesStage  \
-DskipTests -DskipShade \
-DprojectId="$PROJECT" \
-DbucketName="$BUCKET_NAME" \
-DstagePrefix="templates" \
-DtemplateName="Word_Count_Python" \
-pl python -e
```

Reply:
I applied a change to check what's the output (`TemplatePluginUtils.redirectLinesLog`) and the stuckness doesn't happen. I guess there's something funky about `gcloud` when the stream is not being consumed. Pushed that change here.